flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/aggregations.scala

```diff
@@ -414,3 +414,98 @@ case class AggFunctionCall(
     relBuilder.call(this.getSqlAggFunction(), args.map(_.toRexNode): _*)
   }
 }
+
+case class TableAggFunctionCall(
+    aggregateFunction: TableAggregateFunction[_, _],
+    resultTypeInfo: TypeInformation[_],
+    accTypeInfo: TypeInformation[_],
+    args: Seq[Expression],
+    alias: Option[Seq[Expression]],
+    isDistinct: Boolean)
+  extends Aggregation {
+
+  override private[flink] def children: Seq[Expression] = args
+
+  override def resultType: TypeInformation[_] = resultTypeInfo
+
+  override def validateInput(): ValidationResult = {
+    val signature = children.map(_.resultType)
+    // look for a signature that matches the input types
+    val foundSignature = getAccumulateMethodSignature(aggregateFunction, signature)
+    if (foundSignature.isEmpty) {
+      ValidationFailure(s"Given parameters do not match any signature. \n" +
+        s"Actual: ${signatureToString(signature)} \n" +
+        s"Expected: ${
+          getMethodSignatures(aggregateFunction, "accumulate")
+            .map(_.drop(1))
+            .map(signatureToString)
+            .mkString(", ")}")
+    } else {
+      ValidationSuccess
+    }
+  }
+
+  override def toString: String = s"${aggregateFunction.getClass.getSimpleName}($args)"
+
+  override def toAggCall(
+      name: String, isDistinct: Boolean = false)(implicit relBuilder: RelBuilder): AggCall = {
+    relBuilder.aggregateCall(
+      this.getSqlAggFunction(),
+      isDistinct,
+      false,
+      null,
+      name,
+      args.map(_.toRexNode): _*)
+  }
+
+  override private[flink] def getSqlAggFunction()(implicit relBuilder: RelBuilder) = {
+    val typeFactory = relBuilder.getTypeFactory.asInstanceOf[FlinkTypeFactory]
+
+    TableAggSqlFunction(
+      aggregateFunction.functionIdentifier,
+      aggregateFunction.toString,
+      aggregateFunction,
+      alias.map(_.map(_.asInstanceOf[UnresolvedFieldReference].name)),
+      resultType,
+      accTypeInfo,
+      typeFactory)
+  }
+
+  override private[flink] def toRexNode(implicit relBuilder: RelBuilder): RexNode = {
+    relBuilder.call(this.getSqlAggFunction(), args.map(_.toRexNode): _*)
+  }
+}
+
+class TableAggFunctionCallAliasable(
```

Comment:
`TableAggFunctionCallAliasable.as()` returns `TableAggFunctionCall`, so that users can't write multi as, such as `tab.flatAggregate(func('a, 'b) as ('a, 'b, c) as ('d, 'e, 'f))`

Reply:
I meant that move the `as` method into `TableAggFunctionCall`.

Reply:
What's the return type of `as` method? If it returns a `TableAggFunctionCall`. The `TableAggFunctionCall` can still perform another `as`. Users can write query like `tab.flatAggregate(func('a, 'b) as ('a, 'b, c) as ('d, 'e, 'f))`. 
I think we should avoid this invalid case.

Reply:
Hmm,  also make sense to me due to `TableAggFunctionCallAliasable` is inner type. but I'm not sure, because table can be `as` many times. e.g.: `tab.as(..).as(..)`, I think  @twalthr and @fhueske can give us more useful suggestions.